Index: libgringo/gringo/litdep.h
===================================================================
--- libgringo/gringo/litdep.h	(revision 5990)
+++ libgringo/gringo/litdep.h	(working copy)
@@ -111,7 +111,6 @@
 		GrdNodeStack grdStack_;
 		VarNodeVec   varNodes_;
 	};
-}
 
 inline LitDep::GrdNode* new_clone(const LitDep::GrdNode& a)
 {
@@ -119,3 +118,4 @@
 	return 0;
 }
 
+}
Index: libgringo/src/parser.r2c
===================================================================
--- libgringo/src/parser.r2c	(revision 5990)
+++ libgringo/src/parser.r2c	(working copy)
@@ -20,6 +20,7 @@
 #include <gringo/inclit.h>
 #include <gringo/varterm.h>
 #include <gringo/constterm.h>
+#include <gringo/domain.h>
 #include <gringo/prgvisitor.h>
 #include <gringo/predlit.h>
 #include <gringo/rule.h>
Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(revision 5990)
+++ CMakeLists.txt	(working copy)
@@ -58,7 +58,7 @@
 #	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-variable")
 endif()
 
-set(Boost_USE_MULTITHREADED OFF CACHE STRING "Whether to Use multi-threaded boost libraries")
+set(Boost_USE_MULTITHREADED ON)
 find_package(Boost REQUIRED filesystem system)
 include_directories(${Boost_INCLUDE_DIRS})
 if(WITH_LUA STREQUAL "system")
